pgrx/src/datum/array.rs

```diff
@@ -1012,41 +1012,47 @@ where
     }
 }
 
+#[inline]
+/// Converts an iterator into an array datum
+fn array_datum_from_iter<T: IntoDatum>(elements: impl Iterator<Item = T>) -> Option<pg_sys::Datum> {
+    let mut state = unsafe {
+        pg_sys::initArrayResult(
+            T::type_oid(),
+            PgMemoryContexts::CurrentMemoryContext.value(),
+            false,
+        )
+    };
+    for s in elements {
+        let datum = s.into_datum();
+        let isnull = datum.is_none();
+
+        unsafe {
+            state = pg_sys::accumArrayResult(
+                state,
+                datum.unwrap_or(0.into()),
+                isnull,
+                T::type_oid(),
+                PgMemoryContexts::CurrentMemoryContext.value(),
+            );
+        }
+    }
+
+    if state.is_null() {
+        // shouldn't happen
+        None
```

Comment:
I'm probably being overly pedantic because yes, this shouldn't happen.  

So if it does, lets `panic!()` here instead.  Silently returning `None` would be very confusing.

Reply:
Done!

We could also switch to `initArrayResultWithSize` eventually